### PR TITLE
Add torch.unflatten and improve its docs

### PR DESCRIPTION
**Pull request by fdrocha**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81231
* #81230
* __->__ #81399

    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

### CODE REVIEW

File: test/test_overrides.py [hunk 1]
```diff
@@ -649,7 +649,11 @@ def instance_gen():
                     func_args.append(3.5)
                 elif t == 'bool':
                     func_args.append(False)
-                elif t.startswith('int') or t in {'Dimname', 'DimnameList'}:
+                elif t == 'Dimname':
+                    func_args.append("")
+                elif t == 'DimnameList':
+                    func_args.append([""])
+                elif t.startswith('int'):
                     func_args.append(0)
```
Comment (Lezcano) [inline test/test_overrides.py:657]: @peterbell10 does this change sound reasonable?

Comment (peterbell10) [reply test/test_overrides.py:657]: Seems reasonable to me, a dimname should be a string in the python api.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
             [5, 6, 7, 8]])
 """.format(**common_args))
 
+add_docstr(torch.unflatten,
+           r"""
+unflatten(input, dim, sizes) -> Tensor
+
+Expands a dimension of the input tensor over multiple dimensions.
+
+A tensor reshaping function that does the opposite of :func:`torch.flatten`.
```
Comment (Lezcano) [inline torch/_torch_docs.py:4059]: What about using `seealso`?
```suggestion
.. seealso::

    :func:`torch.flatten` the inverse of this function. It coalesces several dimensions into one.
```

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
...
+    {input}
+    dim (int): Dimension to be unflattened, specified as an index into
+         ``input.shape``.
+    sizes (Tuple[int]): New shape of the unflattened dimension.
+         One of its elements can be `-1` in which case the corresponding output
+         dimension is inferred. Otherwise, the product of ``sizes`` *must*
+         equal ``input.shape[dim]``.
+
+Returns:
+    A ``View`` of ``input`` with the specified dimension unflattened.
```
Comment (Lezcano) [inline torch/_torch_docs.py:4071]:
```suggestion
    A view of the input with the specified dimension unflattened.
```

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
...
+
+Examples:
```
Comment (Lezcano) [inline torch/_torch_docs.py:4073]: I don't know why the examples are being rendered funny in the docs. I think you may need to write
```suggestion
Examples::

```
See how it's rendered at the moment (taken from the message from facebook-github-bot) https://docs-preview.pytorch.org/81399/generated/torch.unflatten.html?highlight=unflatten#torch.unflatten

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
...
+    >>> torch.unflatten(torch.randn(3, 4, 1), 1, (2, 2)).shape
+    torch.Size([3, 2, 2, 1])
+    >>> torch.unflatten(torch.randn(3, 4, 1), 1, (-1, 2)).shape
+    torch.Size([3, 2, 2, 1])
+    >>> torch.unflatten(torch.randn(5, 12, 3), -1, (2,2,3,1,1)).shape
```
Comment (Lezcano) [inline torch/_torch_docs.py:4078]:
```suggestion
    >>> torch.unflatten(torch.randn(5, 12, 3), -1, (2, 2, 3, 1, 1)).shape
```

**Review by lezcano — COMMENTED** [review]
A few minor things, but otherwise looks great! Thank you Fabio!

File: test/test_namedtensor.py [hunk 6]
```diff
@@ -1083,22 +1083,13 @@ def test_flatten_nodims(self):
...
-            torch.ones(4).unflatten(0, (['A', 2], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', 2], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', -1], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', 2], ['B', -1])),
```
[the remark this answers is not in the document]
Comment (fdrocha) [reply test/test_namedtensor.py:1101]: Like we discussed, I removed the overload of unflatten that allows you to do stuff like `t.unflatten(0, (['A', -1], ['B',2]))`. You have to do something like `t.unflatten('A', (['A', -1], ['B',2]))` instead. 

This meant that the tests in here had to be changed, and after making the changes there were duplicate tests and I removed them (i.e., `t.unflatten(0, (['A', -1], ['B',2])` and `t.unflatten(-1, (['A', -1], ['B',2])` both became `t.unflatten('A', (['A', -1], ['B',2])` ).

**Review by lezcano — APPROVED** [review]
LGTM

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@fdrocha can you please manually rebase and land?

**Comment by fdrocha** [comment]
This is now failing with

```
The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 

Broken ops: [
	aten::unflatten.int(Tensor(a) self, int dim, int[] sizes, str[]? names=None) -> Tensor(a)
]
```

(Not sure why it wasn't hitting this error before, I guess the tests must have been improved since CI last ran)

This change was intentional and suggested by @Lezcano. The reasoning was that signature was very unusual and it seemed better to replace it with something more regular. Also, the unusual signature made some override tests fail when the free function variant was added. 

The new signature is `(Tensor(a) self, int dim, int[] sizes)` (there's also a pure Dimname variant). 
This means that code like

```
torch.randn(2).unflatten(0, (('B1', -1), ('B2', 1)))
```

will no longer work, but that seems like inconsistent use of Dimnames and ints for specifying dimensions and not clear it should be supported. Note that something like

```
torch.randn(2, names=('A',)).unflatten('A', (('B1', -1), ('B2', 1)))
```

will still work.

Finally, this only affects deprecated Dimname functionality.

@ngimel WDYT?

**Comment by ngimel** [comment]
If we can make this without BC breaking changes we should. While I'm ok removing mixed dimname and number indexing, the complaint is about regular overload, aten::unflatten.int, not the mixed one.

**Comment by fdrocha** [comment]
> If we can make this without BC breaking changes we should. 
We can, we just have to skip some overload tests.
I think main cost is we will end up with a unique type of signature for `torch.unflatten` that doesn't appear anywhere else and might require some special casing in tests.

> While I'm ok removing mixed dimname and number indexing, the complaint is about regular overload, aten::unflatten.int, not the mixed one.

The complaint is about 

```
aten::unflatten.int(Tensor(a) self, int dim, int[] sizes, str[]? names=None) -> Tensor(a)
```

which _is_ the mixed one. If you don't provide the last argument `names` then it's pure int dimensions, the proposed change is removing that last optional argument from the signature so it's no longer mixed.

Anyway, happy to do it either way, let me know what you prefer @ngimel

**Comment by Lezcano** [comment]
This is expected.  That use case was very weird (mixing numbers and dimnames) and it was the only op in PyTorch that had it. Keeping the support for this would have lead to an unnecessarily complicated code, so we dedided to remove it.
Just add it to the relevant skip list.

Fwiw, @ezyang has an open PR trying to remove all dimnames from PyTorch (I really hope he manages to get it merged), so removing thisbparticularly odd overload is not that big a deal IMO

**Comment by fdrocha** [comment]
> This is expected. That use case was very weird (mixing numbers and dimnames) and it was the only op in PyTorch that had it. Keeping the support for this would have lead to an unnecessarily complicated code, so we dedided to remove it. Just add it to the relevant skip list.
> 
> Fwiw, @ezyang has an open PR trying to remove all dimnames from PyTorch (I really hope he manages to get it merged), so removing thisbparticularly odd overload is not that big a deal IMO

Seems the relevant list is `ALLOW_LIST` in `test/forward_backward_compatibility/check_forward_backward_compatibility.py`, does that seem right?

**Comment by ngimel** [comment]
Yes, that's the list the op needs to be added to
